client/state/reducer-registry/test/index.js

```diff
@@ -0,0 +1,391 @@
+	const count = ( state = 1, action ) => ( 'GROW' === action.type ? state + 1 : state );
+
+	const date = ( state = new Date( 0 ), action ) => {
+		switch ( action.type ) {
+			case 'GROW':
+				return new Date( state.getTime() + 1 );
+			case SERIALIZE:
+				return state.getTime();
+			case DESERIALIZE:
+				if ( isValidStateWithSchema( state, schema ) ) {
+					return new Date( state );
+				}
+				return new Date( 0 );
+			default:
+				return state;
+		}
+	};
+	date.hasCustomPersistence = true;
+
+	describe( '#combineReducersAndAddLater', () => {
+		let reducers;
+
+		beforeEach( () => {
+			reducers = combineReducersAndAddLater(
+				{
+					age,
+					height,
+				},
+				'test',
+				Object.create( null )
+			);
+		} );
+
+		const appState = deepFreeze( {
+			age: 20,
+			height: 171,
+		} );
+
+		const partiallyLoadedAppState = deepFreeze( {
+			age: 20,
+			height: {
+				_initialState: 171,
+			},
+		} );
+
+		test( 'should return initial state on init', () => {
+			const state = reducers( undefined, write );
+			expect( state ).to.eql( { age: 0, height: 160 } );
+		} );
+
+		test( 'should not persist height, because it is missing a schema', () => {
+			const state = reducers( appState, write );
+			expect( state ).to.eql( { age: 20, height: 160 } );
+		} );
+
+		test( 'should persist height without a loaded reducer', () => {
+			reducers = combineReducersAndAddLater(
+				{
+					age,
+				},
+				'test',
+				Object.create( null )
+			);
+			const state = reducers( partiallyLoadedAppState, write );
+			expect( state ).to.eql( { age: 20, height: 171 } );
+		} );
+
+		test( 'should not load height, because it is missing a schema', () => {
+			const state = reducers( appState, load );
+			expect( state ).to.eql( { age: 20, height: 160 } );
+		} );
+
+		test( 'should load height into _initialState without a loaded reducer', () => {
+			reducers = combineReducersAndAddLater(
+				{
+					age,
+				},
+				'test',
+				Object.create( null )
+			);
+			const state = reducers( appState, load );
+			expect( state ).to.eql( { age: 20, height: { _initialState: 171 } } );
+		} );
+
+		test( 'should work without any initial reducers', () => {
+			reducers = combineReducersAndAddLater( {}, 'test', Object.create( null ) );
+			const state = reducers( appState, load );
+			expect( state ).to.eql( { age: { _initialState: 20 }, height: { _initialState: 171 } } );
+		} );
+
+		test( 'should validate age', () => {
+			const state = reducers( { age: -5 }, load );
+			expect( state ).to.eql( { age: 0, height: 160 } );
+		} );
+
+		test( 'actions work as expected', () => {
+			const state = reducers( appState, grow );
+			expect( state ).to.eql( { age: 21, height: 172 } );
+		} );
+
+		test( 'nested reducers work on load', () => {
+			const registry = Object.create( null );
+			reducers = combineReducersAndAddLater(
+				{
+					age,
+					height,
+					date,
+				},
+				'person',
+				registry
+			);
+			const nested = combineReducersAndAddLater(
```

Comment:
Would we be nesting `combineReducersAndAddLater` in extensions, or would an inner reducer be calling `combineReducers`?

Reply:
They could call either; the test here ensures both works.